kimchi/src/circuits/expr.rs

```diff
@@ -3149,3 +3149,246 @@ pub mod test {
         assert_eq!(test_4::<Fp, Fp>(Fp::from(5u64)), Fp::from(160u64));
     }
 }
+
+mod quadricization {
+    use super::{ColTrait, Expr, Op2, Variable};
+    use crate::circuits::gate::CurrOrNext;
+    use ark_ff::{One, Zero};
+    use std::{
+        collections::{BTreeMap, HashMap},
+        fmt::Debug,
+        hash::Hash,
+        ops::Neg,
+    };
+
+    ///extension of columns with necessary extra columns
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
+    pub enum ColumnExtended<S: ColTrait + Eq + Hash> {
+        Inner(S),
+        ///extra column used to reduce degree of expressions
+        Extended(usize),
+    }
+    impl<S: ColTrait> ColTrait for ColumnExtended<S> {
+        fn dom(&self) -> super::Domain {
+            match self {
+                ColumnExtended::Inner(s) => s.dom(),
+                ColumnExtended::Extended(_) => super::Domain::D1,
+            }
+        }
+
+        fn latex(&self) -> String {
+            match self {
+                ColumnExtended::Inner(s) => s.latex(),
+                ColumnExtended::Extended(i) => format!("e_{{{i}}}"),
+            }
+        }
+
+        fn text(&self) -> String {
+            match self {
+                ColumnExtended::Inner(s) => s.text(),
+                ColumnExtended::Extended(i) => format!("e[{i}]"),
+            }
+        }
+
+        fn evaluate<F: ark_ff::Field>(
+            self,
+            evals: &crate::proof::ProofEvaluations<crate::proof::PointEvaluations<F>>,
+        ) -> Option<crate::proof::PointEvaluations<F>> {
+            match self {
+                ColumnExtended::Inner(s) => s.evaluate(evals),
+                ///TODO: a change in evals will be required so that it can also provide access to the extended columns
+                ColumnExtended::Extended(e) => todo!(),
+            }
+        }
+
+        fn make_witness(i: usize) -> Self {
+            Self::Inner(S::make_witness(i))
+        }
+
+        fn make_coefficient(i: usize) -> Self {
+            Self::Inner(S::make_coefficient(i))
+        }
+    }
+    ///records expressions that have been extracted into an extra column
+    struct ExprRecorder<C, Col>
+    where
+        C: Hash + Eq,
+        Col: ColTrait + Hash + Eq,
+    {
+        recorded_exprs: HashMap<Expr<C, ColumnExtended<Col>>, usize>,
+        next: usize,
+    }
+
+    impl<C: Hash + Eq, Col: ColTrait + Hash + Eq> ExprRecorder<C, Col> {
+        fn new() -> Self {
+            Self {
+                recorded_exprs: Default::default(),
+                next: 0,
+            }
+        }
+        fn get_id(&mut self, e: Expr<C, ColumnExtended<Col>>) -> usize {
```

Comment:
If I understand the semantics of `ExprRecorder`, it's basically accumulating expressions, to later sort them according to the insertion order. So it's like an ordered hashmap.

In this case, `get_id` is a bit misleading, I'd call it `insert` or `register` instead?